### PR TITLE
ci: Buildkite pipelines as templates

### PR DESCRIPTION
**Pull request by def-**
Automate retrying each job

Symlinks to keep working in the intermediate phase until pipelines are switched.

I don't have permissions for pipelines in Buildkite, so can't change the configurations myself

Fixes: #22093

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by philip-stoev** [comment]
@def- I added you to https://buildkite.com/organizations/materialize/teams/qa/members

**Comment by def-** [comment]
This should be fine to go in. The next step would be to change all the pipeline configs in buildkite to use mkpipeline.py, then a while later remove the symlinks from pipeline.yml to pipeline.template.yml.

File: ci/mkpipeline.py [hunk 1]
```diff
@@ -34,7 +34,7 @@
 from materialize import mzbuild, spawn
 from materialize.mzcompose.composition import Composition
 
-from ..deploy.deploy_util import rust_version
+from .deploy.deploy_util import rust_version
```
Comment (nrainer-materialize) [inline ci/mkpipeline.py:37]: Nit: Why are we using relative imports here? I had thought the linter does not allow that...

**Review by nrainer-materialize — APPROVED** [review]
LGTM

**Comment by def-** [comment]
One problem with this is that I'll need some help from an admin to switch over all the existing pipelines. @philip-stoev Do you  think this is fine or should we only do it for a smaller subset of pipelines?

**Review by philip-stoev — APPROVED** [review]
I think this is fine. Let me sort out the privileges with @benesch and you should be able to go all the way.

File: ci/mkpipeline.py (hunk 1 above)
Comment (def-) [reply ci/mkpipeline.py:37]: I don't know to be honest, also confused me a bit. Didn't want to rock the boat ;)

Comment (nrainer-materialize) [reply ci/mkpipeline.py:37]: That's ok. Ignore it.

**Comment by benesch** [comment]
It would be better if we could maintain backwards compatibility with versions of the repo before this change. Here's how the definition of the tests pipeline looks:

```
steps:
  - command: "if [[ -f ci/test/mkpipeline.sh ]]; then\n  ci/test/mkpipeline.sh\nelse\n  buildkite-agent pipeline upload ci/test/pipeline.yml\nfi"
    label: ":pipeline:"
    priority: 2
    agents:
    - "queue=*"
    - "docker=*"
```

Is it possible to use a similar conditional to only run the `mkpipeline` script if the template for the pipeline exists? For example, something like this would probably work for `deploy_mz`:

```
steps:
  - command: "if [[ -f ci/deploy_mz/pipeline.template.yml ]]; then\n  bin/ci-builder run stable bin/pyactivate -m ci.mkpipeline deploy_mz "$@"\nelse\n  buildkite-agent pipeline upload ci/deploy_mz/pipeline.yml\nfi"
    label: ":pipeline:"
    priority: 2
    agents:
    - "queue=*"
    - "docker=*"
```

That would let you roll out the new templated pipelines immediately, without breaking feature branches that do not include this change.

**Comment by def-** [comment]
Thanks, that's a good idea. Will try it now on all pipelines.

**Pull request merged**